lib/plug/cowboy/drainer.ex

```diff
@@ -0,0 +1,79 @@
+defmodule Plug.Cowboy.Drainer do
+  @moduledoc """
+  TODO: description
+  """
+  use GenServer
+
+  @doc """
+  TODO: description
+
+  ## Options
+
+    * `:id` - The ID for the process.
+      Defaults to `Plug.Cowboy.Drainer`.
+
+    * `:shutdown` - How long to wait for connections to drain.
+      Defaults to 5000ms.
+
+    * `:refs` - The cowboy listener refs to drain connections for.
+      Defaults to all cowboy listeners.
+
+    * `:drain_check_interval` - How frequently to check if a listener's
+      connections have been drained.
+      Defaults to 1000ms.
+  """
+  @spec child_spec(opts :: Keyword.t()) :: Supervisor.child_spec()
+  def child_spec(opts) when is_list(opts) do
+    {spec_opts, opts} = Keyword.split(opts, [:id, :shutdown])
+
+    Supervisor.child_spec(
+      %{
+        id: __MODULE__,
+        start: {__MODULE__, :start_link, [opts]},
+        type: :worker
+      },
+      spec_opts
+    )
+  end
+
+  @doc false
+  def start_link(opts) do
+    GenServer.start_link(__MODULE__, opts)
+  end
+
+  @doc false
+  @impl true
+  def init(opts) do
+    Process.flag(:trap_exit, true)
+    {:ok, opts}
+  end
+
+  @doc false
+  @impl true
+  def terminate(_reason, opts) do
+    opts
+    |> Keyword.get(:refs)
+    |> drain(Keyword.get(opts, :drain_check_interval, 1_000))
+  end
+
+  defp drain(nil, drain_check_interval) do
```

Comment:
There might be an argument for exposing drain on `Plug.Cowboy` to create feature parity with `Plug.Cowboy.https/3` and `Plug.Cowboy.http/3`. However, I don't really understand the use case for when you would use those. So, I'm inclined to leave this here and someone who has a use case for making it available outside the drainer can do that.

Reply:
Yeah, I would wait until there is a use-case.